src/libraries/System.Private.CoreLib/gen/NativeRuntimeEventSourceGenerator.cs

```diff
@@ -0,0 +1,394 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System;
+using System.Collections.Generic;
+using System.IO;
+using System.Linq;
+using System.Xml.Linq;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.Text;
+
+#nullable enable
+
+namespace Generators
+{
+    [Generator]
+    public sealed class NativeRuntimeEventSourceGenerator : IIncrementalGenerator
+    {
+        private static readonly XNamespace EventNs = "http://schemas.microsoft.com/win/2004/08/events";
+
+        public void Initialize(IncrementalGeneratorInitializationContext context)
+        {
+            // Register the generator to run on AdditionalFiles (manifest and inclusion list)
+            IncrementalValuesProvider<AdditionalText> manifestFiles = context.AdditionalTextsProvider.Where(f => f.Path.EndsWith(".man", StringComparison.OrdinalIgnoreCase));
+            IncrementalValuesProvider<AdditionalText> inclusionFiles = context.AdditionalTextsProvider.Where(f => f.Path.EndsWith(".lst", StringComparison.OrdinalIgnoreCase));
+
+            IncrementalValuesProvider<(AdditionalText Left, System.Collections.Immutable.ImmutableArray<AdditionalText> Right)> combined = manifestFiles.Combine(inclusionFiles.Collect());
+
+            context.RegisterSourceOutput(combined, (spc, tuple) =>
+            {
+                AdditionalText manifestFile = tuple.Left;
+                System.Collections.Immutable.ImmutableArray<AdditionalText> inclusionFiles = tuple.Right;
+                string manifestText = manifestFile.GetText(spc.CancellationToken)?.ToString();
+                string inclusionText = inclusionFiles.FirstOrDefault()?.GetText(spc.CancellationToken)?.ToString();
+                if (string.IsNullOrEmpty(manifestText))
+                {
+                    return;
+                }
+
+                var manifest = XDocument.Parse(manifestText);
+                Dictionary<string, HashSet<string>> inclusionList = ParseInclusionListFromString(inclusionText);
+                foreach (KeyValuePair<string, string> kvp in manifestsToGenerate)
+                {
+                    string providerName = kvp.Key;
+                    string className = providerNameToClassNameMap[providerName];
+                    XElement providerNode = manifest.Descendants(EventNs + "provider").FirstOrDefault(e => (string)e.Attribute("name") == providerName);
```

Comment:
Shouldn't this be `XElement?` We are checking for `null` below, but it isn't marked as nullable. I see the `FirstOrDefault()`, so assume that it should be nullable.